### PR TITLE
Generic login + ExtensionProvider login

### PR DESCRIPTION
**Pull request by CiprianDraghici**
### Feature
- Generic login
- Login with Extension Provider

### Contains breaking changes
[x] No

[] Yes

### Updated CHANGELOG
[x] Yes

### Testing
[x] User testing
[] Unit tests

### CODE REVIEW

File: src/core/ProviderFactory.ts [hunk 1]
```diff
@@ -25,18 +29,21 @@ export interface IProviderRecreateFactory extends IProviderFactory {
 
 export enum ProviderTypeEnum {
   iframe = 'iframe',
-  crossWindow = 'crossWindow'
+  crossWindow = 'crossWindow',
```
Comment (arhtudormorar) [inline src/core/ProviderFactory.ts:32]: are we using wallet and crossWindow?

File: src/core/ProviderFactory.ts [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
 import { CrossWindowProvider } from 'lib/sdkWebWalletCrossWindowProvider';
+import { ExtensionProvider } from '@multiversx/sdk-extension-provider';
+import type { IDAppProviderBase } from '@multiversx/sdk-dapp-utils';
 
-export interface IProvider {
-  login: (options?: { token?: string }) => Promise<any>;
+export interface IProvider extends IDAppProviderBase {
+  init: () => Promise<boolean>;
+  login: (options?: { token?: string }) => Promise<string | boolean>;
+  relogin?: () => Promise<void>;
```
Comment (arhtudormorar) [inline src/core/ProviderFactory.ts:8]: why this?

File: src/core/methods/login/helpers/resolveLoginMethod.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { ProviderTypeEnum } from 'core/ProviderFactory';
+import { LoginMethodsEnum } from 'types/enums.types';
+
+export function resolveLoginMethod(providerType: ProviderTypeEnum) {
+  switch (providerType) {
+    case ProviderTypeEnum.iframe:
+      return LoginMethodsEnum.extra;
```
Comment (arhtudormorar) [inline src/core/methods/login/helpers/resolveLoginMethod.ts:7]: do we still use extra?

File: src/core/methods/login/login.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import { IProviderFactory, ProviderFactory } from 'core/ProviderFactory';
+import { NativeAuthConfigType } from 'types/nativeAuth.types';
+import { nativeAuth } from 'services/nativeAuth';
+import { setAddress } from 'store/actions/account';
+import { setLoginMethod, setTokenLogin } from 'store/actions/loginInfo/loginInfoActions';
+import { resolveLoginMethod } from './helpers/resolveLoginMethod';
+import { setAccountProvider } from '../../providers/accountProvider';
+
+export const login = async (config: IProviderFactory, nativeAuthConfig?: NativeAuthConfigType) => {
```
Comment (arhtudormorar) [inline src/core/methods/login/login.ts:9]: use object

File: src/core/methods/login/login.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
+import { IProvider, IProviderFactory, ProviderFactory } from 'core/ProviderFactory';
+import { NativeAuthConfigType } from 'types/nativeAuth.types';
+import { nativeAuth } from 'services/nativeAuth';
+import { setAddress } from 'store/actions/account';
+import { setLoginMethod, setTokenLogin } from 'store/actions/loginInfo/loginInfoActions';
+import { setAccountProvider } from 'core/providers/accountProvider';
+import { getNativeAuthConfig } from 'services/nativeAuth/methods';
+
+async function normalLogin(provider: IProvider) {
```
Comment (arhtudormorar) [inline src/core/methods/login/login.ts:9]: in my oppinion we should have just login with optional params not 2 methods

File: src/core/methods/login/login.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    signature,
+    nativeAuthToken,
+    loginToken,
+    nativeAuthConfig
+  }
+}
+
+export const login = async ({
+  providerConfig,
+  withNativeAuth
```
Comment (arhtudormorar) [inline src/core/methods/login/login.ts:70]: nativeAuth

File: src/core/methods/login/login.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  if(!provider) {
+    throw new Error('Provider not found');
+  }
+
+  await provider.init?.();
+  setAccountProvider(provider);
+  setLoginMethod(providerConfig.type);
+
+  if(withNativeAuth) {
+    if(typeof withNativeAuth === 'boolean') {
```
Comment (arhtudormorar) [inline src/core/methods/login/login.ts:87]: what if it's false?

File: src/core/methods/login/login.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+      withNativeAuth = getNativeAuthConfig(true);
```
Comment (arhtudormorar) [inline src/core/methods/login/login.ts:88]: never reassign props passed to a function, pls add to eslint https://eslint.org/docs/latest/rules/no-param-reassign

File: src/core/methods/login/login.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    const nativeAuthConfig = withNativeAuth as NativeAuthConfigType;
+
+    return await loginWithNativeToken(provider, nativeAuthConfig);
+  } else {
```
Comment (arhtudormorar) [inline src/core/methods/login/login.ts:94]: no need for else

File: src/core/providers/helpers/emptyProvider.ts [hunk 10]
```diff
@@ -106,6 +101,18 @@ export class EmptyProvider implements IDappProvider {
   ping?(): Promise<boolean> {
```
Comment (arhtudormorar) [inline src/core/providers/helpers/emptyProvider.ts:101]: why ping?

File: src/core/providers/helpers/emptyProvider.ts [hunk 11]
```diff
@@ -106,6 +101,18 @@ export class EmptyProvider implements IDappProvider {
...
+
+  setAddress(address: string): IProvider {
+    throw new Error(notInitializedError(`setAddress with address: ${address}`));
+  }
+
+  getAddress(): string | undefined {
+    throw new Error(notInitializedError('getAddress'));
+  }
+
+  getTokenLoginSignature(): string | undefined {
```
Comment (arhtudormorar) [inline src/core/providers/helpers/emptyProvider.ts:113]: TODO: add comment to be removed

File: src/core/providers/helpers/emptyProvider.ts [hunk 12]
```diff
@@ -106,6 +101,18 @@ export class EmptyProvider implements IDappProvider {
   ping?(): Promise<boolean> {
     return Promise.resolve(false);
   }
+
+  setAddress(address: string): IProvider {
```
Comment (arhtudormorar) [inline src/core/providers/helpers/emptyProvider.ts:105]: do we need setAddress?

File: src/store/actions/loginInfo/loginInfoActions.ts [hunk 13]
```diff
@@ -7,7 +7,7 @@ import {
 } from 'store/slices/loginInfo/loginInfo.types';
 import { getStore } from 'store/store';
 
-export const setLoginMethod = (loginMethod: LoginMethodsEnum) =>
+export const setLoginMethod = (loginMethod: LoginMethodsType) =>
```
Comment (arhtudormorar) [inline src/store/actions/loginInfo/loginInfoActions.ts:10]: why change?

File: src/store/slices/loginInfo/loginInfo.types.ts [hunk 14]
```diff
@@ -18,7 +18,7 @@ export interface LoginInfoType {
 }
 
 export interface LoginInfoSliceType {
-  loginMethod: LoginMethodsEnum;
+  loginMethod: LoginMethodsType;
```
Comment (arhtudormorar) [inline src/store/slices/loginInfo/loginInfo.types.ts:21]: it was better as an enum

File: src/core/index.ts [hunk 15]
```diff
@@ -1,2 +1,2 @@
-export * from './ProviderFactory';
+export * from './providers/ProviderFactory';
 export * from './Logger';
```
Comment (arhtudormorar) [inline src/core/index.ts:2]: can be removed

File: src/core/methods/login/login.ts [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+import {
+  IProvider,
+  IProviderFactory
+} from 'core/providers/types/providerFactory.types';
+import { ProviderFactory } from 'core/providers/ProviderFactory';
+import { tokenLoginSelector } from 'store/selectors';
+import { getState } from 'store/store';
+import { NativeAuthConfigType } from 'services/nativeAuth/nativeAuth.types';
+import { isLoggedIn } from 'utils/account/isLoggedIn';
+import { getAddress } from '../../../utils/account/getAddress';
```
Comment (arhtudormorar) [inline src/core/methods/login/login.ts]: absolute?

File: src/core/methods/login/login.ts [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+  setAddress(address);
+
+  return {
+    address
+  };
+}
+
+async function loginWithNativeToken(
+  provider: IProvider,
+  nativeAuthConfig: NativeAuthConfigType
```
Comment (arhtudormorar) [inline src/core/methods/login/login.ts:37]: why not read from store?

File: src/core/methods/login/login.ts [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    nativeAuthConfig
+  };
+}
+
+export const login = async ({
+  providerConfig
+}: {
+  providerConfig: IProviderFactory;
+}) => {
+  const loggedIn = isLoggedIn();
```
Comment (arhtudormorar) [inline src/core/methods/login/login.ts:86]: getIsLoggedIn()

File: src/core/methods/login/login.ts [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+
+  if (!provider) {
+    throw new Error('Provider not found');
+  }
+
+  await provider.init?.();
+  setAccountProvider(provider);
+  setProviderType(providerConfig.type);
+
+  const nativeAuthConfig = tokenLoginSelector(getState())?.nativeAuthConfig;
```
Comment (arhtudormorar) [inline src/core/methods/login/login.ts:104]: nativeAuthConfigSelector ?

File: src/core/methods/login/login.ts [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+  IProviderFactory
+} from 'core/providers/types/providerFactory.types';
+import { ProviderFactory } from 'core/providers/ProviderFactory';
+import { tokenLoginSelector } from 'store/selectors';
+import { getState } from 'store/store';
+import { NativeAuthConfigType } from 'services/nativeAuth/nativeAuth.types';
+import { isLoggedIn } from 'utils/account/isLoggedIn';
+import { getAddress } from '../../../utils/account/getAddress';
+
+async function normalLogin(provider: IProvider) {
```
Comment (arhtudormorar) [inline src/core/methods/login/login.ts:19]: rename to `plainLogin` ?

File: src/core/providers/ProviderFactory.ts [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+        break;
+    }
+
+    return createdProvider;
+  }
+
+  private async getCrossWindowProvider({
+    walletAddress
+  }: Partial<IProviderConfig['network']>) {
+    // CrossWindowProvider.getInstance().clearInstance();
```
Comment (arhtudormorar) [inline src/core/providers/ProviderFactory.ts:68]: comment ?

File: src/core/providers/accountProvider.ts [hunk 22]
```diff
@@ -1,8 +1,8 @@
-import { IDappProvider } from 'types/dappProvider.types';
 import { emptyProvider } from './helpers/emptyProvider';
 import { CrossWindowProvider } from 'lib/sdkWebWalletCrossWindowProvider';
+import { IProvider } from 'core/providers/types/providerFactory.types';
 
-export type ProvidersType = IDappProvider | CrossWindowProvider;
+export type ProvidersType = IProvider | CrossWindowProvider;
```
Comment (arhtudormorar) [inline src/core/providers/accountProvider.ts:5]: is the | needed ?

File: src/services/nativeAuth/nativeAuth.ts [hunk 23]
```diff
@@ -23,7 +23,7 @@ export const nativeAuth = (config?: NativeAuthConfigType) => {
     extraInfo: extraInfoFromConfig,
     gatewayUrl,
     extraRequestHeaders
-  } = getNativeAuthConfig(config) as NativeAuthConfigType;
+  } = buildNativeAuthConfig(config) as NativeAuthConfigType;
```
Comment (arhtudormorar) [inline src/services/nativeAuth/nativeAuth.ts:26]: casting should not be needed

File: src/store/actions/loginInfo/loginInfoActions.ts [hunk 24]
```diff
@@ -24,6 +25,20 @@ export const setTokenLoginSignature = (signature: string) =>
     }
   });
 
+export const setTokenLoginNativeAuthTokenConfig = (
+  nativeAuthConfig: NativeAuthConfigType
+) =>
+  getStore().setState(({ loginInfo: state }) => {
+    if (state?.tokenLogin != null) {
+      state.tokenLogin.nativeAuthConfig = nativeAuthConfig;
```
Comment (arhtudormorar) [inline src/store/actions/loginInfo/loginInfoActions.ts:33]: return

File: src/utils/account/isLoggedIn.ts [hunk 25]
```diff
@@ -0,0 +1,5 @@
+import { getAddress } from './getAddress';
+
+export const isLoggedIn = () => {
```
Comment (arhtudormorar) [inline src/utils/account/isLoggedIn.ts:3]: getIsLoggedIn

File: src/core/providers/ProviderFactory.ts [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+  ProviderTypeEnum
+} from './types/providerFactory.types';
+
+export class ProviderFactory {
+  public async create({
+    type,
+    config,
+    customProvider
+  }: IProviderFactory): Promise<IProvider | undefined> {
+    let createdProvider: IProvider | undefined = undefined;
```
Comment (razvantomegea) [inline src/core/providers/ProviderFactory.ts:16]: No need o initialise with undefined

File: src/core/providers/ProviderFactory.ts [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+
+    switch (type) {
+      // case ProviderTypeEnum.iframe: {
```
Comment (razvantomegea) [inline src/core/providers/ProviderFactory.ts:19]: WIP?

File: src/core/providers/types/providerFactory.types.ts [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+    walletAddress: string;
+  };
+}
+
+export enum ProviderTypeEnum {
+  iframe = 'iframe',
+  crossWindow = 'crossWindow',
+  extension = 'extension',
+  walletConnect = 'walletConnect',
+  hardware = 'hardware',
```
Comment (razvantomegea) [inline src/core/providers/types/providerFactory.types.ts:26]: hardware means ledger or any other cold wallet?

File: src/store/actions/loginInfo/loginInfoActions.ts (hunk 24 above)
Comment (razvantomegea) [reply src/store/actions/loginInfo/loginInfoActions.ts:33]: `return state`, to be more precise

**Review by razvantomegea — CHANGES_REQUESTED** [review]

File: src/core/providers/ProviderFactory.ts (hunk 27 above)
Comment (CiprianDraghici) [reply src/core/providers/ProviderFactory.ts:19]: yes

File: src/core/providers/types/providerFactory.types.ts (hunk 28 above)
Comment (CiprianDraghici) [reply src/core/providers/types/providerFactory.types.ts:26]: we support only ledger, but I kept a generic name for future integrations

File: src/core/methods/login/login.ts (hunk 20 above)
Comment (CiprianDraghici) [reply src/core/methods/login/login.ts:19]: no, totally disagree

File: src/core/providers/ProviderFactory.ts (hunk 21 above)
Comment (CiprianDraghici) [reply src/core/providers/ProviderFactory.ts:68]: wip, will keep this

File: src/core/providers/accountProvider.ts (hunk 22 above)
Comment (CiprianDraghici) [reply src/core/providers/accountProvider.ts:5]: yes...at least until the next PR when webWalletLogin will be refactored in the same way as extension provider login

**Review by arhtudormorar — APPROVED** [review]

**Review by razvantomegea — APPROVED** [review]

**Pull request merged**